Data/Primitive/PrimRef.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE RoleAnnotations #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE Unsafe #-}
+
+-- | Note: Edward Kmett wrote everything in this module. 
+module Data.Primitive.PrimVar
+  ( 
+  -- * Primitive References
+    PrimVar(..)
+  , newPrimVar
+  , newPinnedPrimVar
+  , newAlignedPinnedPrimVar
+  , readPrimVar
+  , writePrimVar
+  , primVarContents
+  , primVarToMutablePrimArray
+  , casInt
+  , fetchAddInt
+  , fetchSubInt
+  , fetchAndInt
+  , fetchNandInt
+  , fetchOrInt
+  , fetchXorInt
+  , atomicReadInt
+  , atomicWriteInt
+  ) where
+
+import Control.Monad.Primitive
+import Data.Primitive
+import GHC.Exts
+import GHC.Ptr (castPtr)
+
+--------------------------------------------------------------------------------
+-- * Primitive References
+--------------------------------------------------------------------------------
+
+newtype PrimVar s a = PrimVar (MutablePrimArray s a)
+
+type role PrimVar nominal nominal
+
```

Comment:
```suggestion
```
I'm not sure this is necessary, at least the other types in this package don't have any role annotations. The `s` seems to be nominal anyway, while `a` is phantom by default.

Reply:
Shouldn't the other types have role annotations? Doing this makes it safe, `Prim` cannot read out of bounds because of a type change.

Reply:
Hmm, that's a good point. So representational should be fine for `a`, but we can also make it nominal for now and then relax it later.

Reply:
I don't think representational would be fine, because in theory you could still give a different implementation of `Prim`.

Reply:
We could also have a `unsafeCoercePrimVar` if a coercion is needed.

Reply:
Having the role be explicit is important. The inferred role (phantom) allows user to coerce a `PrimVar s Word8` to a `PrimVar s Word64`, which is unsound. `PrimArray` really should have a role annotation as well.